packages/payment-detection/src/erc20/escrow-info-retriever.ts

```diff
@@ -0,0 +1,125 @@
+type EscrowArgs = {
+  paymentReference: string;
+};
+
+/**
+ * Retrieves a list of payment events from a payment reference, a destination address, a token address and a escrow contract.
+ */
+export default class EscrowERC20InfoRetriever
+  implements PaymentTypes.IPaymentNetworkBaseInfoRetriever<PaymentTypes.IPaymentNetworkBaseEvent> {
+  public contractEscrow: ethers.Contract;
+  public provider: ethers.providers.Provider;
+
+  /**
+   * @param paymentReference The reference to identify the payment.
+   * @param escrowContractAddress The address of the escrow contract.
+   * @param escrowCreationBlockNumber The block that created the escrow contract.
+   * @param network The Ethereum network to use.
+   */
+  constructor(
+    private paymentReference: string,
+    private escrowContractAddress: string,
+    private escrowCreationBlockNumber: number,
+    private network: string,
+  ) {
+    // Creates a local or default provider
+    this.provider = getDefaultProvider(this.network);
+
+    // Setup the ERC20 escrow contract interface
+    this.contractEscrow = new ethers.Contract(
+      this.escrowContractAddress,
+      erc20EscrowContractAbiFragment,
+      this.provider,
+    );
+  }
+
+  /**
+   * Retrieves events for the current contract, address and network.
+   */
+  public async getContractEvents(): Promise<PaymentTypes.IPaymentNetworkBaseEvent[]> {
+    // Create a filter to find all the RequestFrozen logs with the payment reference
+    const freezeFilter = this.contractEscrow.filters.RequestFrozen(
+      '0x' + this.paymentReference,
+    ) as ethers.providers.Filter;
+    freezeFilter.fromBlock = this.escrowCreationBlockNumber;
+    freezeFilter.toBlock = 'latest';
+
+    // Create a filter to find all the Init Emergency logs with the payment reference
+    const initEmergencyFilter = this.contractEscrow.filters.InitiatedEmergencyClaim(
+      '0x' + this.paymentReference,
+    ) as ethers.providers.Filter;
+    freezeFilter.fromBlock = this.escrowCreationBlockNumber;
+    freezeFilter.toBlock = 'latest';
+
+    // Create a filter to find all the Fee Transfer logs with the payment reference
+    const revertEmergencyFilter = this.contractEscrow.filters.RevertedEmergencyClaim(
+      '0x' + this.paymentReference,
+    ) as ethers.providers.Filter;
+    freezeFilter.fromBlock = this.escrowCreationBlockNumber;
+    freezeFilter.toBlock = 'latest';
+
+    // Get the RequestFrozen event logs
+    const freezeLogs = await this.provider.getLogs(freezeFilter);
+
+    // Get the InitiateEmergencyClaim event logs
+    const initEmergencyLogs = await this.provider.getLogs(initEmergencyFilter);
+
+    // Get the RequestFrozen event logs
+    const revertEmergencyLogs = await this.provider.getLogs(revertEmergencyFilter);
+
+    interface EthersLogsWithEventName extends ethers.providers.Log {
+      eventName: PaymentTypes.EVENTS_NAMES;
+    }
+
+    // Merge events if multiple logs
+    const logs: EthersLogsWithEventName[] = [
```

Comment:
I think it would be easier to understand if the mapping with `eventName` was done separately, for instance when instanciating each array of logs.